qiskit/transpiler/preset_passmanagers/level3.py

```diff
@@ -197,53 +198,57 @@ def _opt_control(property_set):
+            target,
+            basis_gates,
+            approximation_degree,
+            unitary_synthesis_method,
+            unitary_synthesis_plugin_config,
+        )
     init.append(RemoveResetInZeroState())
     init.append(OptimizeSwapBeforeMeasure())
     init.append(RemoveDiagonalGatesBeforeMeasure())
     if coupling_map or initial_layout:
-        layout = PassManager()
-        layout.append(_given_layout)
-        layout.append(_choose_layout_0, condition=_choose_layout_condition)
-        layout.append(_choose_layout_1, condition=_vf2_match_not_found)
-        layout += common.generate_embed_passmanager(coupling_map)
-        vf2_call_limit = None
-        if pass_manager_config.layout_method is None and pass_manager_config.initial_layout is None:
-            vf2_call_limit = int(3e7)  # Set call limit to ~60 sec with retworkx 0.10.2
-        routing = common.generate_routing_passmanager(
-            routing_pass,
-            target,
-            coupling_map=coupling_map,
-            vf2_call_limit=vf2_call_limit,
-            backend_properties=backend_properties,
-            seed_transpiler=seed_transpiler,
-            use_barrier_before_measurement=not toqm_pass,
-        )
+        if layout_method not in {"trivial", "dense", "noise_adaptive", "sabre"}:
+            layout = plugin_manager.get_passmanager_stage(
+                "layout", layout_method, pass_manager_config
+            )
+        else:
+            layout = PassManager()
+            layout.append(_given_layout)
+            layout.append(_choose_layout_0, condition=_choose_layout_condition)
+            layout.append(_choose_layout_1, condition=_vf2_match_not_found)
+            layout += common.generate_embed_passmanager(coupling_map)
+        routing = routing_pm
     else:
         layout = None
         routing = None
-    translation = common.generate_translation_passmanager(
-        target,
-        basis_gates,
-        translation_method,
-        approximation_degree,
-        coupling_map,
-        backend_properties,
-        unitary_synthesis_method,
-        unitary_synthesis_plugin_config,
-    )
+    if translation_method not in {"translator", "synthesis", "unroller"}:
+        translation = plugin_manager.get_passmanager_stage(
+            "translation", translation_method, pass_manager_config
+        )
+    else:
+        translation = common.generate_translation_passmanager(
+            target,
+            basis_gates,
+            translation_method,
+            approximation_degree,
+            coupling_map,
+            backend_properties,
+            unitary_synthesis_method,
+            unitary_synthesis_plugin_config,
+        )
     pre_routing = None
     if toqm_pass:
         pre_routing = translation
-    optimization = PassManager()
-    unroll = [pass_ for x in translation.passes() for pass_ in x["passes"]]
-    optimization.append(_depth_check + _size_check)
+    optimization = None
```

Comment:
This line is not in level 1 and 2. Probably for a good reason I could not find.

Reply:
It's because of the nested construction of the pre-optimization stage as part of optimization (which is needed because level 3 does extra optimizations and might need to run gate direction again and the construction is based on whether the target is used or not). I'll expand out this logic to make it easier to follow

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/8305/commits/4190774c3b50a1a9b7832437dd4492f8ddea4bfc